### PR TITLE
Multi-VC support - Part 1

### PR DESCRIPTION
**Pull request by rohitjogvmw**
Code changes for  below changes

 - Send hostname from each node. 
 - If config file is not present then Initialize WorkerNode 
 - Modified gcfg config file read struct to make use of new format 
 - CloudProvider calls each provider when node is registered / unregistered

Testing done:
   - After creating kubernetes cluster on vsphere, manually changed hostname and VMName.
     After rebooting workernode, I could see that new hostname node got registered
   - For new node addition, on master verified that NodeRegistered interfaced got called
   - On one of the worker node modified  /etc/systemd/system/multi-user.target.wants/kubelet.service and removed cloud-config parameter to kubectl and restarted kubelet. Node was shown in ReadyState once kubelet was started.

Below is the sample output after all above changes :

```
kubectl get nodes
NAME          STATUS                     AGE       VERSION
master        Ready,SchedulingDisabled   1d        v1.6.0-alpha.0.13043+2211945f3cad85
node2         Ready                      1d        v1.6.0-alpha.0.13043+2211945f3cad85
node3         Ready                      1d        v1.6.0-alpha.0.13043+2211945f3cad85
node4         Ready                      1d        v1.6.0-alpha.0.13043+2211945f3cad85
rohit-node1   Ready                      18m       v1.6.0-alpha.0.13046+389d9886d96fd4
```

@divyenpatel, @luomiao, @tusharnt, @BaluDontu, @SandeepPissay

### CODE REVIEW

File: pkg/cloudprovider/providers/vsphere/vsphere.go [hunk 1]
```diff
@@ -299,47 +313,12 @@ func (vs *VSphere) getVMByName(ctx context.Context, nodeName k8stypes.NodeName)
 // NodeAddresses is an implementation of Instances.NodeAddresses.
 func (vs *VSphere) NodeAddresses(nodeName k8stypes.NodeName) ([]v1.NodeAddress, error) {
 	// Get local IP addresses if node is local node
-	if vs.localInstanceID == nodeNameToVMName(nodeName) {
+	if vs.hostName == nodeNameToVMName(nodeName) {
 		return getLocalIP()
 	}
-	addrs := []v1.NodeAddress{}
```
Comment (divyenpatel) [inline pkg/cloudprovider/providers/vsphere/vsphere.go]: Why is the code block L305:L342 removed?
What is the output of `kubectl describe node node1 | grep Address`
```
$ kubectl describe node node1 | grep Address
Addresses:		10.162.12.189,10.162.12.189,node1
```
This should not return internal addresses and ipv6 addresses

File: pkg/cloudprovider/providers/vsphere/vsphere.go [hunk 2]
```diff
@@ -746,3 +706,13 @@ func (vs *VSphere) DeleteVolume(vmDiskPath string) error {
 	vclib.RecordvSphereMetric(vclib.OperationDeleteVolume, requestTime, err)
 	return err
 }
+
+// Notification handler when node is registered.
+func (vs *VSphere) NodeRegistered(node *v1.Node) {
+	glog.Errorf("Node Registered: %+v", node)
```
Comment (divyenpatel) [inline pkg/cloudprovider/providers/vsphere/vsphere.go]: glog.Errorf => glog.V(4).Infof

File: pkg/cloudprovider/providers/vsphere/vsphere.go [hunk 3]
```diff
@@ -746,3 +706,13 @@ func (vs *VSphere) DeleteVolume(vmDiskPath string) error {
...
+}
+
+// Notification handler when node is unregistered.
+func (vs *VSphere) NodeUnregistered(node *v1.Node) {
+	glog.Errorf("Node Unregistered: %+v", node)
```
Comment (divyenpatel) [inline pkg/cloudprovider/providers/vsphere/vsphere.go]: glog.Errorf => glog.V(4).Infof

File: pkg/cloudprovider/providers/vsphere/vsphere.go [hunk 4]
```diff
@@ -151,19 +168,38 @@ func readConfig(config io.Reader) (VSphereConfig, error) {
...
+		return nil, err
+	}
+
+	return &vs, nil
+}
+
+// Creates new Contreoller node interface and returns
+func newControlllerNode(cfg VSphereConfig) (*VSphere, error) {
+	var err error
+	glog.Errorf("cfg is\n%+v\n", cfg)
```
Comment (divyenpatel) [inline pkg/cloudprovider/providers/vsphere/vsphere.go]: glog.Errorf => glog.V(4).Infof

File: pkg/cloudprovider/providers/vsphere/vsphere.go [hunk 5]
```diff
@@ -100,6 +100,23 @@ type VSphereConfig struct {
...
+		// vCenter username.
+		User string `gcfg:"user"`
+		// vCenter password in clear text.
+		Password string `gcfg:"password"`
+		// vCenter port.
+		VCenterPort string `gcfg:"port"`
+		// True if vCenter uses self-signed cert.
+		InsecureFlag bool `gcfg:"insecure-flag"`
+		// Datacenter in which VMs are located.
+		Datacenter string `gcfg:"datacenter"`
```
Comment (BaluDontu) [inline pkg/cloudprovider/providers/vsphere/vsphere.go]: Shouldn't this Datacenters? User can specify multiple datacenters within each VirtualCenter.

File: pkg/cloudprovider/providers/vsphere/vsphere.go [hunk 6]
```diff
@@ -100,6 +100,23 @@ type VSphereConfig struct {
...
+		// WorkingDir is path where VMs can be found.
+		WorkingDir string `gcfg:"working-dir"`
```
Comment (BaluDontu) [inline pkg/cloudprovider/providers/vsphere/vsphere.go]: Shouldn't WorkingDir should be in the Global struct?

File: pkg/cloudprovider/providers/vsphere/vsphere.go [hunk 7]
```diff
@@ -100,6 +100,23 @@ type VSphereConfig struct {
 		VMName string `gcfg:"vm-name"`
 	}
```
Comment (BaluDontu) [inline pkg/cloudprovider/providers/vsphere/vsphere.go]: I guess you will remove all the unnecessary fields in Global fields later on? For example VMUUID, VMName etc. is unnecessary.

Remove VCenterIP, modify Datacenter -> Datacenters.

Comment (BaluDontu) [inline pkg/cloudprovider/providers/vsphere/vsphere.go]: Can you rename Datastore -> DefaultDatastore??

File: pkg/cloudprovider/providers/vsphere/vsphere.go (hunk 5 above)
Comment (BaluDontu) [inline pkg/cloudprovider/providers/vsphere/vsphere.go]: Also how to identify which VirtualCenter map entry is master? I think we should add a field to specify which of the VC entry is a master? Otherwise we would not know it.

File: pkg/cloudprovider/providers/vsphere/vsphere.go [hunk 8]
```diff
@@ -151,19 +168,38 @@ func readConfig(config io.Reader) (VSphereConfig, error) {
...
+	if err != nil {
+		glog.Errorf("Failed to get hostname. err: %+v", err)
+		return nil, err
+	}
+
+	return &vs, nil
+}
+
+// Creates new Contreoller node interface and returns
+func newControlllerNode(cfg VSphereConfig) (*VSphere, error) {
```
Comment (BaluDontu) [inline pkg/cloudprovider/providers/vsphere/vsphere.go]: typo in function name. newControlllerNode -> newControllerNode.

File: pkg/cloudprovider/providers/vsphere/vsphere.go (hunk 4 above)
Comment (BaluDontu) [inline pkg/cloudprovider/providers/vsphere/vsphere.go]: U don't need to specify /n at the end of log. By default the next log entry will be in a new line. Please add some spaces between the log, otherwise it is difficult to read.

File: pkg/cloudprovider/providers/vsphere/vsphere.go (hunk 1 above)
Comment (BaluDontu) [inline pkg/cloudprovider/providers/vsphere/vsphere.go]: +1. Didn't understand why is this removed?

File: pkg/cloudprovider/providers/vsphere/vsphere.go (hunk 7 above)
Comment (rohitjogvmw) [inline pkg/cloudprovider/providers/vsphere/vsphere.go]: Balu, during my next task ready new format config file I will remove whatever is not needed.

Changed Datacenter to Datacenters.
Changed Datastore to DefaultDatastore.

File: pkg/cloudprovider/providers/vsphere/vsphere.go (hunk 5 above)
Comment (rohitjogvmw) [inline pkg/cloudprovider/providers/vsphere/vsphere.go]: Why do we need master VC?

File: pkg/cloudprovider/providers/vsphere/vsphere.go (hunk 6 above)
Comment (rohitjogvmw) [inline pkg/cloudprovider/providers/vsphere/vsphere.go]: I think user may want us to use different workingDir on different VCs.
If on all VCs workingDir is same, then user can specify only once in Global.

File: pkg/cloudprovider/providers/vsphere/vsphere.go (hunk 8 above)
Comment (rohitjogvmw) [inline pkg/cloudprovider/providers/vsphere/vsphere.go]: fixed.

File: pkg/cloudprovider/providers/vsphere/vsphere.go (hunk 4 above)
Comment (rohitjogvmw) [inline pkg/cloudprovider/providers/vsphere/vsphere.go]: I have just kept it for now. We can't log cfg anyways as it will print VC credentials in logs which we don't want.
I have added TODO comment to remove it once everything is settled.

File: pkg/cloudprovider/providers/vsphere/vsphere.go (hunk 1 above)
Comment (rohitjogvmw) [inline pkg/cloudprovider/providers/vsphere/vsphere.go]: On workerNode, there are no VC credentials so there is no VC connection.

Comment (rohitjogvmw) [inline pkg/cloudprovider/providers/vsphere/vsphere.go]: Objective of this task is to report hostName and not VMName.
Since VMName is not needed to report there is no need to connect to VC and generate VMName

Comment (rohitjogvmw) [inline pkg/cloudprovider/providers/vsphere/vsphere.go]: I observed that each kubelet reports it's own localIP and we actually don't need to find out addresses of some OtherNode.

kubectl describe node node1
shows correct IP addresses.

During testing I even has IPv6 enabled VMs IPv6 addresses were correctly filtered out.

File: pkg/cloudprovider/providers/vsphere/vsphere.go (hunk 2 above)
Comment (rohitjogvmw) [inline pkg/cloudprovider/providers/vsphere/vsphere.go]: fixed.

**Comment by divyenpatel** [comment]
LGTM

File: pkg/cloudprovider/providers/vsphere/vsphere.go (hunk 6 above)
Comment (BaluDontu) [inline pkg/cloudprovider/providers/vsphere/vsphere.go]: No workingDir is only used to create the DummyVM by the master. workingDir will only be useful and required by the master VM. As per my understanding, we don't need workingDir on all VC's.

File: pkg/cloudprovider/providers/vsphere/vsphere.go (hunk 5 above)
Comment (BaluDontu) [inline pkg/cloudprovider/providers/vsphere/vsphere.go]: Ok, i get it now. We scan all the VC's to identify which VC does this VM belong to. In that case, it would be not needed.
